src/main/webapp/app/overview/course-exercises/course-exercises-grouped-by-category.component.ts

```diff
@@ -0,0 +1,147 @@
+            exerciseGroup.isCollapsed = false;
+        });
+    }
+
+    private restoreStateBeforeSearch(exerciseGroups: ExerciseGroups) {
+        this.searchWasActive = false;
+
+        Object.entries(exerciseGroups).forEach(([exerciseGroupKey, exerciseGroup]) => {
+            exerciseGroup.isCollapsed = this.exerciseGroupsBeforeSearch[exerciseGroupKey].isCollapsed;
+        });
+    }
+
+    private keepCurrentCollapsedOrExpandedStateOfExerciseGroups(exerciseGroups: ExerciseGroups) {
+        Object.entries(exerciseGroups).forEach(([exerciseGroupKey, exerciseGroup]) => {
+            exerciseGroup.isCollapsed = this.exerciseGroups[exerciseGroupKey].isCollapsed;
+        });
+    }
+
+    private makeSureAtLeastOneExerciseIsExpanded(exerciseGroups: ExerciseGroups) {
+        const exerciseGroupsWithExercises = Object.entries(exerciseGroups).filter(([, exerciseGroup]) => exerciseGroup.exercises.length > 0);
+        const expandedExerciseGroups = exerciseGroupsWithExercises.filter(([, exerciseGroup]) => !exerciseGroup.isCollapsed);
+
+        const atLeastOneExerciseIsExpanded = expandedExerciseGroups.length > 0;
+        if (!atLeastOneExerciseIsExpanded && exerciseGroupsWithExercises.length > 0) {
+            exerciseGroupsWithExercises[0][1].isCollapsed = false;
+        }
+    }
+
+    /**
+     * 1. Expand all sections with matches on search
+     * 2. Keep the expanded or collapsed state of the exercise groups when a filter is applied
+     * 3. Make sure at least one displayed section is expanded by default
+     *
+     * @param exerciseGroups updated and grouped exercises that are to be displayed
+     */
+    private adjustExpandedOrCollapsedStateOfExerciseGroups(exerciseGroups: ExerciseGroups) {
+        const isSearchingExercise = this.appliedSearchString;
+        if (isSearchingExercise) {
+            return this.expandAllExercisesAndSaveStateBeforeSearch(exerciseGroups);
+        }
+
+        if (this.searchWasActive) {
+            return this.restoreStateBeforeSearch(exerciseGroups);
+        }
+
+        const filterIsApplied = this.exerciseGroups;
+        if (filterIsApplied) {
+            this.keepCurrentCollapsedOrExpandedStateOfExerciseGroups(exerciseGroups);
+        }
+
+        this.makeSureAtLeastOneExerciseIsExpanded(exerciseGroups);
+    }
+
+    private getExerciseGroup(exercise: Exercise): ExerciseGroupCategory {
+        if (!exercise.dueDate) {
+            return 'noDueDate';
+        }
+
+        const dueDate = dayjs(exercise.dueDate);
+        const now = dayjs();
+
+        const dueDateIsInThePast = dueDate.isBefore(now);
+        if (dueDateIsInThePast) {
+            return 'previous';
+        }
+
+        const dueDateIsWithinNextWeek = dueDate.isBefore(now.add(1, 'week'));
+        if (dueDateIsWithinNextWeek) {
+            return 'current';
+        }
+
+        return 'future';
+    }
+
+    protected readonly Object = Object;
```

Comment:
Usually, these are defined at the top of the class

Reply:
Some changes were needed anyways - I've moved it to the top now